coderd/subdomain.go

```diff
@@ -0,0 +1,152 @@
+package coderd
+
+import (
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/coder/coder/coderd/httpmw"
+
+	"github.com/go-chi/chi/v5"
+
+	"golang.org/x/xerrors"
+)
+
+const (
+	// XForwardedHostHeader is a header used by proxies to indicate the
+	// original host of the request.
+	XForwardedHostHeader = "X-Forwarded-Host"
+)
+
+// ApplicationURL is a parsed application url into it's components
+type ApplicationURL struct {
+	AppName       string
+	WorkspaceName string
+	Agent         string
+	Username      string
+	Path          string
+	Domain        string
+}
+
+func (api *API) handleSubdomain(middlewares ...func(http.Handler) http.Handler) func(http.Handler) http.Handler {
```

Comment:
Seems like this should go in `httpmw` instead to follow existing patterns. Is there a reason it can't be that I'm missing?

Reply:
Yea, the organization of this is a bit annoying.

The bottom calls `api.proxyWorkspaceApplication`. So the subdomain and path based code all use the app proxy code. If it's in `httpmw`, then I have to export that function and pass it into the `httpmw` function somehow?

That seemed a bit more annoying to read.